array_api_tests/hypothesis_helpers.py

```diff
@@ -55,17 +58,27 @@ def mutually_promotable_dtypes(draw, dtype_objects=dtype_objects):
     # pairs (XXX: Can we redesign the strategies so that they can prefer
     # shrinking dtypes over values?)
     sorted_table = sorted(promotion_table)
-    sorted_table = sorted(sorted_table, key=lambda ij: -1 if ij[0] == ij[1] else sorted_table.index(ij))
-    dtype_pairs = [(dtype_mapping[i], dtype_mapping[j]) for i, j in
-                   sorted_table]
-
-    filtered_dtype_pairs = [(i, j) for i, j in dtype_pairs if i in
-                            dtype_objects and j in dtype_objects]
+    sorted_table = sorted(
+        sorted_table, key=lambda ij: -1 if ij[0] == ij[1] else sorted_table.index(ij)
+    )
+    dtype_pairs = [(dtype_mapping[i], dtype_mapping[j]) for i, j in sorted_table]
     if FILTER_UNDEFINED_DTYPES:
-        filtered_dtype_pairs = [(i, j) for i, j in filtered_dtype_pairs
-                                if not isinstance(i, _UndefinedStub)
-                                and not isinstance(j, _UndefinedStub)]
-    return draw(sampled_from(filtered_dtype_pairs))
+        dtype_pairs = [(i, j) for i, j in dtype_pairs
+                       if not isinstance(i, _UndefinedStub)
+                       and not isinstance(j, _UndefinedStub)]
+    return dtype_pairs
+
+def promotable_dtypes(dtype):
```

Comment:
Drawing a dtype then using promotable_dtypes will have a different sampling distribution than mutually_promotable_dtype_pairs. I don't know if that matters. The asymmetry of `promotable_dtypes` does feel a little weird. I think we might want to instead try to use mutually_promotable_dtypes in the elementwise strategies so that they are clearer.

Reply:
Makes sense, fixed!

Just noting that if you could figure out a different place for `dtype_mapping` and `promotion_table`, you could remove the otherwise-unnecessary composite wrapper for `mutually_promotable_dtype_pairs` (see the comment I wrote).

Reply:
The various data bits in the promotion test file should indeed be refactored and cleaned up. It's already used in quite a few places outside of that file. It's something I intend to do eventually, but if you see something that can help then go ahead. 

I think I also may have made a mistake by using the string values of the dtypes as the main keys in the dictionaries rather than the dtype objects themselves, leading to a lot of code where you have multiple dict lookups to get at what you want. My only concern is that the dtype objects themselves might not actually be hashable, but I think this isn't a problem in practice, so we perhaps shouldn't worry about it.

Reply:
I opened https://github.com/data-apis/array-api-tests/issues/21 for this.